### ISSUE TITLE
Worker timeout when requesting compare/runs/<x>...<y>, 502 Bad Gateway seen by user agent

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#599]
Seen by Antoine: 502 Bad Gateway response when GETting 
`conbench.ursa.dev/compare/runs/ef44eed05b61439ebfaf377a482d5f49...2d61d83ce5814bea96e47ebd04a625fe/`
`conbench.ursa.dev/compare/runs/a02a6dc9375543fea2f9a003f3bb341f...5c32ae93b4dd47ec854cb84ff040a770/`

Reproduced. Web application log around the time of request shows

```
Jan 17 15:39:29.254 i-0d9f4f0020ae49a35 conbench [2023-01-17 14:39:28 +0000] [389] [INFO] Booting worker with pid: 389
Jan 17 15:39:28.254 i-0d9f4f0020ae49a35 conbench [2023-01-17 14:39:27 +0000] [386] [INFO] Worker exiting (pid: 386)
Jan 17 15:39:28.254 i-0d9f4f0020ae49a35 conbench [2023-01-17 14:39:27 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:386)
```

We saw two symptoms, and can assess potential causes:

- The logs (datadog) show gunicorn worker timeouts around the time when requesting this compare view. That's probably because the worker process that is supposed to generate an HTTP response to the request doesn't do so within the deadline (30 seconds). In that case gunicorn seemingly does not generate an HTTP response at all, and terminates the worker process.

- The 502 Bad Gateway is then emitted from a reverse proxy (probably nginx, as part of k8s ingress controller) because the upstream (nginx term) probably terminates the TCP connection before it has sent a response.

I think there are two root issues:

- The HTTP response to those compare/runs/.... takes way too long to generate because of 'a lot of work to do'.
- There is no insight (via logs) about which part of that work takes very long.

We should explore why the response generation taskes so long (add logging, etc), and then envision architectural changes to make the compare/runs view 10x or 100x faster :).

**Comment by jonkeane** [comment]
Separately from figuring out the root cause + ensuring that we rectify that. Is it possible to also add alerts (possibly in datadog?) when 502s?

**Comment by austin3dickey** [comment]
I think it's likely that my history query PR added to the workload for requests like this. See discussion here: https://github.com/conbench/conbench/pull/507#issuecomment-1359892065

**Comment by jgehrcke** [comment]
> Separately from figuring out the root cause + ensuring that we rectify that. Is it possible to also add alerts (possibly in datadog?) when 502s?

Yes, the canonical setup here is to have the system emit basic metrics (https://github.com/conbench/conbench/issues/493), to measure the rate of specific errors, and then do alerts based on error rates.

**Comment by austin3dickey** [comment]
Sorry this comment starts to ramble long, but I'm writing all my thoughts down as I investigate.

I'm exploring the API calls that happen when you load that page. I set up a read-only connection to the database from my laptop, and am calling the same code that runs when you GET https://conbench.ursa.dev/compare/runs/ef44eed05b61439ebfaf377a482d5f49...2d61d83ce5814bea96e47ebd04a625fe.

(The code that runs is [this method](https://github.com/conbench/conbench/blob/f7a68b6e7cbcea9d77204e47b80ec54ccdaef4ad/conbench/app/compare.py#L141).)

1. We GET `/api/compare/runs/ef44eed05b61439ebfaf377a482d5f49...2d61d83ce5814bea96e47ebd04a625fe`. On my laptop this takes 11 seconds to return comparisons for all 3332 applicable BenchmarkResults.
2. GET `/api/runs/ef44eed05b61439ebfaf377a482d5f49` (<1 sec)
3. GET `/api/runs/2d61d83ce5814bea96e47ebd04a625fe` (<1 sec)
4. GET `/api/benchmarks/run_id=2d61d83ce5814bea96e47ebd04a625fe` (5 sec for 3153 results)
5. GET `/api/history/919975f04bbd4320bbbedc8e9d1b153f` (<1 sec for 96 results)
6. GET `/api/history/cd23a9e0417e4ce6b8e3f096626cd79f` (<1 sec for 96 results)
7. GET `/api/history/8707ab296d8a46aa95b22aab229c8f00` (<1 sec for 96 results)

The total time is around 17 seconds. Of course, it takes longer in production, timing out after 30. Notably, when I just try to do the first API call (`/api/compare`) in my web browser or using `curl`, it takes from 20 seconds to more than 30 (when it times out). So I think that's the main problem. 

(It is kinda weird that we get 3000ish results twice. I'm not sure what exactly we need from `/api/benchmarks/` that we don't currently get from `/api/compare/runs/`. If we got rid of number 4 up above, it could speed things up a little.)

Digging deeper, most of the time in (1) and (4) is spent in [set_z_scores()](https://github.com/conbench/conbench/blob/f7a68b6e7cbcea9d77204e47b80ec54ccdaef4ad/conbench/entities/distribution.py#L157). But what's interesting is that every time we call the function, the caches perform optimally: we only run the meat of the function on the first BenchmarkResult and hit the two caches on every subsequent BenchmarkResult.

Looking further, we don't even spend that much time on the meaty database queries like I thought! `get_baseline_run()` and the Distribution query are fast. We seem to be spending most of the time on this line, repeating it for every iteration:

https://github.com/conbench/conbench/blob/f7a68b6e7cbcea9d77204e47b80ec54ccdaef4ad/conbench/entities/distribution.py#L217-L223

Wow, that's a bad data structure for this use case. This should be a lookup dict. My bad!

***

Okay, I restructured this code so we're not doing something like a O(n^2) lookup every time we call `set_z_scores()`. Locally, it cuts the time of my test script in half. I'm excited to submit the PR and see if it helps the timeout issue. We could always try to restructure how we construct the Compare webapp page later, but I think this improvement to `set_z_scores()` will be a great first step.
